src/controller/CHIPDeviceController.cpp

```diff
@@ -1178,6 +1184,298 @@ void DeviceCommissioner::OnSessionEstablished()
+        // TODO: Map error status to correct error code
+        commissioner->OnSessionEstablishmentError(CHIP_ERROR_INTERNAL);
+    }
+}
+
+CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(chip::ByteSpan attestationElements, chip::ByteSpan signature)
+{
+    VerifyOrReturnError(mState == State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mDeviceBeingPaired < kNumMaxActiveDevices, CHIP_ERROR_INCORRECT_STATE);
+
+    Device * device = &mActiveDevices[mDeviceBeingPaired];
+
+    ByteSpan certDeclaration;
+    ByteSpan attestationNonce;
+    ByteSpan timestamp;
+    ByteSpan firmwareInfo;
+    ByteSpan vendorReserved1;
+    ByteSpan vendorReserved2;
+    ByteSpan vendorReserved3;
+    ByteSpan vendorReserved4;
+
+    ByteSpan * element_array[] = { &certDeclaration, &attestationNonce, &timestamp,       &firmwareInfo,
+                                   &vendorReserved1, &vendorReserved2,  &vendorReserved3, &vendorReserved4 };
+
+    System::PacketBufferTLVWriter tlvWriter;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+    System::PacketBufferHandle attestationElementsTbs;
+
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+    int i                      = 0;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+    do
+    {
+        if (!TLV::IsContextTag(tlvReader.GetTag()))
+        {
+            continue;
+        }
+        const uint8_t * data = nullptr;
+        ReturnErrorOnFailure(tlvReader.GetDataPtr(data));
+        *element_array[i++] = ByteSpan(data, tlvReader.GetLength());
+    } while (tlvReader.Next() == CHIP_NO_ERROR);
+
+    // Step f
+    // Retrieve attestation challenge
+    PeerConnectionState * state = mSessionMgr->GetPeerConnectionState(
+        { mPairingSession.PeerConnection().GetPeerNodeId(), mPairingSession.GetPeerKeyId(), mAdminId });
+    VerifyOrReturnError(state != nullptr, CHIP_ERROR_NOT_CONNECTED);
+    ByteSpan attestationChallenge = state->GetSecureSession().GetAttestationChallenge();
+
+    // Create attestation-elements-tbs
+    attestationElementsTbs = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrReturnError(!attestationElementsTbs.IsNull(), CHIP_ERROR_NO_MEMORY);
+
+    tlvWriter.Init(std::move(attestationElementsTbs));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(0), attestationElements));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), attestationChallenge));
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize(&attestationElementsTbs));
```

Comment:
Same comment as for the generation in AttestationResponse: this is not TLV per spec, and no allocation is necessary. Hash for signature verification can be computed incrementally with stream version of SHA256.

Reply:
fixed